File: instrumentation/jetty-jmx-7/src/main/java/org/eclipse/jetty/server/Server_Instrumentation.java
/*
 *
 *  * Copyright 2020 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package org.eclipse.jetty.server;

import com.newrelic.agent.bridge.AgentBridge;
import com.newrelic.api.agent.NewRelic;
import com.newrelic.api.agent.weaver.NewField;
import com.newrelic.api.agent.weaver.Weave;
import com.newrelic.api.agent.weaver.Weaver;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;

@Weave(originalName = "org.eclipse.jetty.server.Server")
public class Server_Instrumentation {

    @NewField
    private static final AtomicBoolean addedJmx = new AtomicBoolean(false);

    protected void doStart() throws Exception {
        if (!addedJmx.getAndSet(true)) {
            AgentBridge.jmxApi.addJmxMBeanGroup("org.eclipse.jetty");
            NewRelic.getAgent().getLogger().log(Level.FINER, "Added JMX for Jetty");
        }
        Weaver.callOriginal();
    }

}

File: instrumentation/play-2.6.13/src/main/scala/play/core/server/Server_Instrumentation.scala
/*
 *
 *  * Copyright 2020 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package play.core.server

import com.newrelic.api.agent.weaver.Weaver
import com.newrelic.api.agent.weaver.scala.{ScalaMatchType, ScalaWeave}
import com.newrelic.api.agent.{NewRelic, Trace}
import com.nr.agent.instrumentation.play26.PlayUtil
import play.api.mvc.{Handler, RequestHeader, Result}
import play.core.ApplicationProvider

import scala.concurrent._;

@ScalaWeave(`type` = ScalaMatchType.Object, `originalName` = "play.core.server.Server")
class Server_Instrumentation {
  @Trace(dispatcher = true)
  def getHandlerFor(
    request: RequestHeader,
    applicationProvider: ApplicationProvider
  ): Either[Future[Result], (RequestHeader, Handler)] = {
    val tempResult: Either[Future[Result], (RequestHeader, Handler)] = Weaver.callOriginal()
    val result = tempResult.right.map { case (request, handler) =>
      // In order to correctly track async work from here we need to stash a token into the request.
      val token = NewRelic.getAgent.getTransaction.getToken
      (PlayUtil.appendToken(request, token), handler)
    }

    val tracer = NewRelic.getAgent.getTracedMethod
    if (tracer != null) {
      tracer.setMetricName("Play2Routing");
    }

    result
  }

}
